// calcite-react/Alert/Alert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _AlertStyled = require("./Alert-styled");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Alert = function Alert(_ref) {
  var children = _ref.children,
      closeLabel = _ref.closeLabel,
      onClose = _ref.onClose,
      other = _objectWithoutProperties(_ref, ["children", "closeLabel", "onClose"]);

  var getAlertClose = function getAlertClose(closeLabel) {
    if (closeLabel) {
      return _react.default.createElement(_AlertStyled.StyledAlertClose, {
        onClick: onClose
      }, closeLabel);
    }
  };

  return _react.default.createElement(_AlertStyled.StyledAlert, other, children, getAlertClose(closeLabel));
};

Alert.propTypes = {
  /** Components to be rendered within the Alert. */
  children: _propTypes.default.node,

  /** Color modifier for the Alert. */
  blue: _propTypes.default.bool,

  /** Color modifier for the Alert. */
  green: _propTypes.default.bool,

  /** Color modifier for the Alert. */
  yellow: _propTypes.default.bool,

  /** Color modifier for the Alert. */
  red: _propTypes.default.bool,

  /** Full-width modifier for the Alert. */
  full: _propTypes.default.bool,

  /** Display label used to close the Alert. */
  closeLabel: _propTypes.default.node,

  /** Callback function fired when the close link is clicked. */
  onClose: _propTypes.default.func
};
Alert.defaultProps = {
  onClose: function onClose() {}
};
Alert.displayName = 'Alert';
var _default = Alert;
exports.default = _default;
