// calcite-react/Accordion/Accordion-styled.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyledChevronIcon = exports.StyledAccordionContent = exports.StyledAccordionSection = exports.StyledAccordionTitle = exports.StyledAccordion = void 0;

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _helpers = require("../utils/helpers");

var _SideNavStyled = require("../SideNav/SideNav-styled");

var _ChevronRightIcon = _interopRequireDefault(require("calcite-ui-icons-react/ChevronRightIcon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

// Copyright 2019 Esri
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.​
var StyledAccordion = (0, _styledComponents.default)(_SideNavStyled.StyledSideNav).withConfig({
  displayName: "StyledAccordion",
  componentId: "sc-16ghmal-0"
})(["border-radius:", ";"], function (props) {
  return props.theme.borderRadius;
});
exports.StyledAccordion = StyledAccordion;

var StyledAccordionSection = _styledComponents.default.div.withConfig({
  displayName: "StyledAccordionSection",
  componentId: "sc-16ghmal-1"
})([""]);

exports.StyledAccordionSection = StyledAccordionSection;
var StyledAccordionTitle = (0, _styledComponents.default)(_SideNavStyled.StyledSideNavTitle).withConfig({
  displayName: "StyledAccordionTitle",
  componentId: "sc-16ghmal-2"
})(["display:flex;align-items:center;cursor:pointer;", ":first-child &{border-top-left-radius:", ";border-top-right-radius:", ";}", ":last-child &{border-bottom-left-radius:", ";border-bottom-right-radius:", ";}&:hover,&:focus{transition:all,", ";background-color:", ";outline:none;}", ";"], StyledAccordionSection, function (props) {
  return props.theme.borderRadius;
}, function (props) {
  return props.theme.borderRadius;
}, StyledAccordionSection, function (props) {
  return props.theme.borderRadius;
}, function (props) {
  return props.theme.borderRadius;
}, (0, _helpers.transition)(), function (props) {
  return props.theme.palette.lightestGray;
}, function (props) {
  return props.disabled && (0, _styledComponents.css)(["opacity:0.5;pointer-events:none;"]);
});
exports.StyledAccordionTitle = StyledAccordionTitle;

var StyledAccordionContent = _styledComponents.default.div.withConfig({
  displayName: "StyledAccordionContent",
  componentId: "sc-16ghmal-3"
})(["display:none;", ";"], function (props) {
  return props.active && (0, _styledComponents.css)(["display:block;"]);
});

exports.StyledAccordionContent = StyledAccordionContent;
var StyledChevronIcon = (0, _styledComponents.default)(_ChevronRightIcon.default).withConfig({
  displayName: "StyledChevronIcon",
  componentId: "sc-16ghmal-4"
})(["width:20;height:20;transition:transform ", ";", ";"], (0, _helpers.transition)(), function (props) {
  return props.active === 'true' && (0, _styledComponents.css)(["transform:rotate(90deg);"]);
});
exports.StyledChevronIcon = StyledChevronIcon;
