// calcite-react/Toaster/doc/ToasterExampleComponent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _Button = _interopRequireDefault(require("../../Button"));

var _Tooltip = _interopRequireDefault(require("../../Tooltip"));

var _InformationIcon = _interopRequireDefault(require("calcite-ui-icons-react/InformationIcon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Framework and third-party non-ui
var StyledToasterExample = _styledComponents.default.div.withConfig({
  displayName: "StyledToasterExample",
  componentId: "sc-19zkpmy-0"
})(["display:flex;align-items:center;justify-content:space-between;"]);

var ToasterExampleComponent = function ToasterExampleComponent(props) {
  return _react.default.createElement(StyledToasterExample, null, _react.default.createElement("span", null, "A Toaster Was Added"), _react.default.createElement(_Tooltip.default, {
    title: "More Info",
    placement: "left"
  }, _react.default.createElement(_Button.default, {
    onClick: props.onInfoClick,
    type: "button",
    iconButton: true,
    icon: _react.default.createElement(_InformationIcon.default, null)
  })));
};

var _default = ToasterExampleComponent;
exports.default = _default;
