// calcite-react/SubNav/SubNav-styled.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyledMultiRowActions = exports.StyledSubNavActions = exports.StyledSubNavTitle = exports.StyledSubNavList = exports.StyledSubNavLink = exports.StyledSubNavLeftContent = exports.StyledSubNav = void 0;

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _Elements = require("../Elements");

var _helpers = require("../utils/helpers");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

// Copyright 2019 Esri
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.​
var StyledSubNav = _styledComponents.default.header.withConfig({
  displayName: "StyledSubNav",
  componentId: "rv6lg3-0"
})(["position:relative;background-color:", ";display:flex;flex-wrap:wrap;", ";", ";"], function (props) {
  return props.theme.palette.lightestGray;
}, function (props) {
  return props.blue && (0, _styledComponents.css)(["background-color:", ";"], props.theme.palette.darkerBlue);
}, function (props) {
  return props.backgroundImage && (0, _styledComponents.css)(["", ";"], (0, _helpers.backgroundGradient)(props.backgroundImage, props.gradientFromColor, props.gradientToColor, props.overlayGradient));
});

exports.StyledSubNav = StyledSubNav;

var StyledSubNavLeftContent = _styledComponents.default.div.withConfig({
  displayName: "StyledSubNavLeftContent",
  componentId: "rv6lg3-1"
})(["display:flex;flex:1 0 200px;justify-content:space-between;flex-direction:column;"]);

exports.StyledSubNavLeftContent = StyledSubNavLeftContent;
var StyledSubNavLink = (0, _styledComponents.default)(_Elements.CalciteA).withConfig({
  displayName: "StyledSubNavLink",
  componentId: "rv6lg3-2"
})(["padding:0.25em 0.75em;margin:0 0 0 0.25em;font-family:", ";color:", ";", ";background-color:", ";box-sizing:border-box;border-radius:", " ", " 0 0;transition:background-color ", ",color ", ";display:inline-block;&:hover,&:focus{background-color:", ";color:", ";text-decoration:none;", ";}", " .active > &,.active > &:hover,.active > &:focus{background-color:", ";color:", ";}", ";"], function (props) {
  return props.theme.avenirFamily;
}, function (props) {
  return props.theme.palette.offWhite;
}, (0, _helpers.fontSize)(-1), function (props) {
  return props.theme.palette.transparentOffBlack;
}, function (props) {
  return props.theme.borderRadius;
}, function (props) {
  return props.theme.borderRadius;
}, (0, _helpers.transition)(), (0, _helpers.transition)('150ms'), function (props) {
  return props.theme.palette.transparentDarkerGray;
}, function (props) {
  return props.theme.palette.white;
}, function (props) {
  return (0, _helpers.subNavUnderline)(props);
}, function (props) {
  return props.active && (0, _styledComponents.css)(["&,&:hover,&:focus{background-color:", ";color:", ";}"], props.theme.palette.white, props.theme.palette.offBlack);
}, function (props) {
  return props.theme.palette.white;
}, function (props) {
  return props.theme.palette.offBlack;
}, function (props) {
  return props.disabled && (0, _styledComponents.css)(["opacity:0.5;pointer-events:none;"]);
});
exports.StyledSubNavLink = StyledSubNavLink;

var StyledSubNavList = _styledComponents.default.nav.withConfig({
  displayName: "StyledSubNavList",
  componentId: "rv6lg3-3"
})(["display:flex;height:32px;margin-top:0em;padding-left:0.25em;box-sizing:border-box;"]);

exports.StyledSubNavList = StyledSubNavList;

var StyledSubNavActions = _styledComponents.default.nav.withConfig({
  displayName: "StyledSubNavActions",
  componentId: "rv6lg3-4"
})(["margin:0.5em;display:flex;justify-content:flex-end;align-items:center;"]);

exports.StyledSubNavActions = StyledSubNavActions;
var StyledSubNavTitle = (0, _styledComponents.default)(_Elements.CalciteH1).withConfig({
  displayName: "StyledSubNavTitle",
  componentId: "rv6lg3-5"
})(["", ";padding-left:0.25em;margin-top:", ";margin-bottom:", ";line-height:1.25 box-sizing:border-box;", ";"], (0, _helpers.fontSize)(4), function (props) {
  return (0, _helpers.unitCalc)(props.theme.baseline, 2, '/');
}, function (props) {
  return (0, _helpers.unitCalc)(props.theme.baseline, 2, '/');
}, function (props) {
  return props.blue && (0, _styledComponents.css)(["color:", ";"], props.theme.palette.white);
});
exports.StyledSubNavTitle = StyledSubNavTitle;

var StyledMultiRowActions = _styledComponents.default.div.withConfig({
  displayName: "StyledMultiRowActions",
  componentId: "rv6lg3-6"
})(["display:flex;flex-direction:column;"]);

exports.StyledMultiRowActions = StyledMultiRowActions;
