// calcite-react/Table/Table.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableContext = exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _TableStyled = require("./Table-styled");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var TableContext = (0, _react.createContext)({
  tableContext: {
    blue: undefined,
    striped: undefined,
    plain: undefined,
    noTable: undefined,
    justified: undefined,
    noCol: undefined,
    noRow: undefined
  }
});
exports.TableContext = TableContext;

var Table = function Table(_ref) {
  var children = _ref.children,
      blue = _ref.blue,
      striped = _ref.striped,
      plain = _ref.plain,
      noTable = _ref.noTable,
      justified = _ref.justified,
      noCol = _ref.noCol,
      noRow = _ref.noRow,
      other = _objectWithoutProperties(_ref, ["children", "blue", "striped", "plain", "noTable", "justified", "noCol", "noRow"]);

  var tableContext = {
    blue: blue,
    striped: striped,
    plain: plain,
    noTable: noTable,
    justified: justified,
    noCol: noCol,
    noRow: noRow
  };
  return _react.default.createElement(TableContext.Provider, {
    value: {
      tableContext: tableContext
    }
  }, _react.default.createElement(_TableStyled.StyledTable, _extends({
    blue: blue,
    plain: plain,
    noTable: noTable,
    noCol: noCol,
    noRow: noRow
  }, other), children));
};

exports.default = Table;
Table.propTypes = {
  /** The content of the component. */
  children: _propTypes.default.node,

  /** A style prop to render a blue Table. */
  blue: _propTypes.default.bool,

  /** A style prop to render a Table with striped rows. */
  striped: _propTypes.default.bool,

  /** A style prop to render Table with no borders or background color. */
  plain: _propTypes.default.bool,

  /** A style prop to render Table with no styling. */
  noTable: _propTypes.default.bool,

  /** A style prop to remove leading and trailing padding on a Table. */
  justified: _propTypes.default.bool,

  /** A style prop to render a Table with no column borders. */
  noCol: _propTypes.default.bool,

  /** A style prop to render a Table with no row borders. */
  noRow: _propTypes.default.bool
};
Table.defaultProps = {};
Table.displayName = 'Table';
